### ISSUE TITLE
Unused variables analysis is order dependent

### ISSUE THREAD

**Issue by mchf** [YorickPeterse/ruby-lint#195]
Lets have a class definition:
```
# UnusedInstanceVariable class
class UnusedInstanceVariable
  def init_variable
    @variable = false
  end

  def result
    res &&= @variable != true
    puts "UnusedInstanceVariable result: #{res}"
  end
end
```

then ```ruby-lint``` will pass with no problem. However, if we swap method definitions this way:
```
# UnusedInstanceVariable class
class UnusedInstanceVariable
  def result
    res &&= @variable != true
    puts "UnusedInstanceVariable result: #{res}"
  end

  def init_variable
    @variable = false
  end
end
```

then the analysis reports:
```
unused_instance_variable.rb: warning: line 9, column 5: unused instance variable @variable
```

It behaves the same way even if the variable is initialized in class scope.

**Comment by YorickPeterse** [comment]
This is because ruby-lint processes things from top to bottom as it goes along, meaning that when `@variable` is used ruby-lint is not yet aware of its existence. Sadly there's not much that can be done about this without a substantial rewrite.

See the following links for more information: 

* https://github.com/YorickPeterse/ruby-lint/issues/109#issuecomment-39129114
* https://github.com/YorickPeterse/ruby-lint/issues/156#issuecomment-150010920
